File: pkg/waylog/v2/parity/parity_test.go
package parity

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"strings"
	"testing"
	"time"

	waylogv2 "github.com/sssmaran/WaylogCLI/pkg/waylog/v2"
)

// scenarioRunner carries per-scenario plumbing: a fresh SDK Init, the
// emit buffer the SDK writes to, and the *eventv2.Event captured from
// the relevant Finalize* call.
type scenarioRunner struct {
	t   *testing.T
	buf *bytes.Buffer
}

func newScenario(t *testing.T) *scenarioRunner {
	t.Helper()
	if err := drainSDK(t); err != nil {
		t.Fatalf("pre-test drain: %v", err)
	}
	buf := &bytes.Buffer{}
	cfg := waylogv2.Config{
		Service: "checkout",
		Env:     "test",
		Output:  buf,
	}
	if err := waylogv2.Init(cfg); err != nil {
		t.Fatalf("Init: %v", err)
	}
	return &scenarioRunner{t: t, buf: buf}
}

// drainSDK uses the public API to wait for in-flight requests to finish
// from any prior scenario. It is the public-API-only equivalent of
// resetForTest().
func drainSDK(t *testing.T) error {
	t.Helper()
	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
	defer cancel()
	return waylogv2.Shutdown(ctx)
}

// emitted parses the most recent JSON line written to the SDK output.
// All scenarios produce exactly one final event.
func (r *scenarioRunner) emitted() []byte {
	r.t.Helper()
	out := r.buf.Bytes()
	if len(out) == 0 {
		r.t.Fatal("no event emitted")
	}
	lines := bytes.Split(bytes.TrimRight(out, "\n"), []byte{'\n'})
	return lines[len(lines)-1]
}

// assertParity compares the SDK's emitted JSON against the named fixture
// after masking non-deterministic fields. On mismatch it prints both
// blobs side by side so the diff is human-readable in test output.
func (r *scenarioRunner) assertParity(name string) {
	r.t.Helper()
	got := r.emitted()
	want, err := LoadFixture(name)
	if err != nil {
		r.t.Fatalf("load fixture %s: %v", name, err)
	}
	eq, err := Equal(got, want)
	if err != nil {
		r.t.Fatalf("compare: %v", err)
	}
	if !eq {
		gotPretty := jsonPretty(r.t, got)
		wantPretty := jsonPretty(r.t, want)
		gotMasked := jsonPretty(r.t, mustMask(r.t, got))
		wantMasked := jsonPretty(r.t, mustMask(r.t, want))
		r.t.Fatalf("parity mismatch for %s\n--- emitted (raw) ---\n%s\n--- fixture (raw) ---\n%s\n--- emitted (masked) ---\n%s\n--- fixture (masked) ---\n%s",
			name, gotPretty, wantPretty, gotMasked, wantMasked)
	}
}

func mustMask(t *testing.T, raw []byte) []byte {
	t.Helper()
	out, err := MaskNonDeterministic(raw)
	if err != nil {
		t.Fatalf("mask: %v", err)
	}
	return out
}

func jsonPretty(t *testing.T, raw []byte) string {
	t.Helper()
	var v any
	if err := json.Unmarshal(raw, &v); err != nil {
		return string(raw)
	}
	out, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return string(raw)
	}
	return string(out)
}

// --- Scenario 1: ok-simple --------------------------------------------------

func TestParity_OKSimple(t *testing.T) {
	r := newScenario(t)
	ctx := waylogv2.Begin(context.Background(), waylogv2.BeginOptions{
		TraceID: strings.Repeat("1", 32),
		SpanID:  strings.Repeat("1", 16),
	})
	waylogv2.SetField(ctx, "http", map[string]any{
		"method": "POST",
		"route":  "/checkout",
		"status": 200,
	})
	if err := waylogv2.StepVoid(ctx, "db.load_cart", func(ctx context.Context) error {
		return nil
	}); err != nil {
		t.Fatalf("step: %v", err)
	}
	if _, err := waylogv2.Finalize(ctx); err != nil {
		t.Fatalf("Finalize: %v", err)
	}
	r.assertParity("ok-simple.json")
}

// --- Scenario 2: error-payment-cascade --------------------------------------

func TestParity_ErrorPaymentCascade(t *testing.T) {
	r := newScenario(t)
	ctx := waylogv2.Begin(context.Background(), waylogv2.BeginOptions{
		TraceID: strings.Repeat("2", 32),
		SpanID:  strings.Repeat("2", 16),
	})
	waylogv2.SetField(ctx, "http", map[string]any{
		"method": "POST",
		"route":  "/checkout",
		"status": 200, // overwritten by SetHTTPStatus(502) below
	})
	waylogv2.SetField(ctx, "user", map[string]any{"id": "u_123"})

	if err := waylogv2.StepVoid(ctx, "db.load_cart", func(ctx context.Context) error {
		return nil
	}); err != nil {
		t.Fatalf("db step: %v", err)
	}

	wantErr := waylogv2.NewError("PMT_502", waylogv2.WithReason("upstream gateway 5xx"))
	stepErr := waylogv2.StepVoid(ctx, "payment.charge", func(ctx context.Context) error {
		waylogv2.From(ctx).Warn("retrying payment", nil)
		waylogv2.RecordOutgoingSpan(ctx, strings.Repeat("3", 16), "payment", "POST /charge")
		waylogv2.Fail(ctx, wantErr)
		return wantErr
	})
	if !errors.Is(stepErr, wantErr) {
		t.Fatalf("expected payment.charge to surface %v, got %v", wantErr, stepErr)
	}

	waylogv2.SetHTTPStatus(ctx, 502)

	if _, err := waylogv2.Finalize(ctx); err != nil {
		t.Fatalf("Finalize: %v", err)
	}
	r.assertParity("error-payment-cascade.json")
}

// --- Scenario 3: error-panic ------------------------------------------------

func TestParity_ErrorPanic(t *testing.T) {
	r := newScenario(t)
	ctx := waylogv2.Begin(context.Background(), waylogv2.BeginOptions{
		TraceID: strings.Repeat("6", 32),
		SpanID:  strings.Repeat("7", 16),
	})
	waylogv2.SetField(ctx, "http", map[string]any{
		"method": "POST",
		"route":  "/checkout",
		"status": 500,
	})
	if _, err := waylogv2.FinalizePanic(ctx); err != nil {
		t.Fatalf("FinalizePanic: %v", err)
	}
	r.assertParity("error-panic.json")
}

// --- Scenario 4: suppressed-healthcheck -------------------------------------

func TestParity_SuppressedHealthcheck(t *testing.T) {
	r := newScenario(t)
	ctx := waylogv2.Begin(context.Background(), waylogv2.BeginOptions{
		TraceID: strings.Repeat("5", 32),
		SpanID:  strings.Repeat("6", 16),
	})
	waylogv2.SetField(ctx, "http", map[string]any{
		"method": "GET",
		"route":  "/healthz",
		"status": 200,
	})
	waylogv2.Suppress(ctx)
	if _, err := waylogv2.Finalize(ctx); err != nil {
		t.Fatalf("Finalize: %v", err)
	}
	r.assertParity("suppressed-healthcheck.json")
}

// --- Scenario 5: aborted-cancel ---------------------------------------------

func TestParity_AbortedCancel(t *testing.T) {
	r := newScenario(t)
	parent, cancel := context.WithCancel(context.Background())
	ctx := waylogv2.Begin(parent, waylogv2.BeginOptions{
		TraceID: strings.Repeat("4", 32),
		SpanID:  strings.Repeat("5", 16),
	})
	waylogv2.SetField(ctx, "http", map[string]any{
		"method": "POST",
		"route":  "/checkout",
	})
	cancel()
	if _, err := waylogv2.FinalizeAborted(ctx); err != nil {
		t.Fatalf("FinalizeAborted: %v", err)
	}
	r.assertParity("aborted-cancel.json")
}

// --- Scenario 6: timeout-watchdog -------------------------------------------
//
// The fixture encodes a watchdog firing while payment.charge is still
// active: the previous step closed normally, but the watchdog wins the
// seal before payment.charge can return. Reproducing that via public API
// means firing FinalizeTimeout from inside the active step's fn so the
// step is still on the stack at seal time. The SDK's flush-active-steps
// path then snapshots payment.charge into steps[] as the closed
// status-ok entry the fixture expects.

func TestParity_TimeoutWatchdog(t *testing.T) {
	r := newScenario(t)
	ctx := waylogv2.Begin(context.Background(), waylogv2.BeginOptions{
		TraceID: strings.Repeat("3", 32),
		SpanID:  strings.Repeat("4", 16),
	})
	waylogv2.SetField(ctx, "http", map[string]any{
		"method": "POST",
		"route":  "/checkout",
	})

	if err := waylogv2.StepVoid(ctx, "db.load_cart", func(ctx context.Context) error {
		return nil
	}); err != nil {
		t.Fatalf("db.load_cart step: %v", err)
	}

	if err := waylogv2.StepVoid(ctx, "payment.charge", func(ctx context.Context) error {
		// Fire the watchdog while still inside the step. Once Finalize
		// seals the request, the surrounding StepVoid wrapper sees the
		// sealed flag and skips its addStep — but the lifecycle path has
		// already snapshotted payment.charge into steps[].
		if _, err := waylogv2.FinalizeTimeout(ctx); err != nil {
			return err
		}
		return nil
	}); err != nil {
		t.Fatalf("payment.charge step: %v", err)
	}

	r.assertParity("timeout-watchdog.json")
}

File: packages/waylog-ts/src/__tests__/parity.test.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import { begin, fail, finalize, finalizeAborted, finalizePanic, finalizeTimeout, from, init, newError, recordOutgoingSpan, setField, setHTTPStatus, stepSync, suppress } from "../index.js";
import type { WideEvent } from "../types.js";

const fixtureDir = join(process.cwd(), "..", "..", "testdata", "fixtures", "v2");

function capture() {
  const lines: string[] = [];
  init({ service: "checkout", env: "test", output: (line) => lines.push(line) });
  return {
    event(): WideEvent {
      const raw = lines.find((line) => line.trim().startsWith("{"));
      if (!raw) throw new Error("no event");
      return JSON.parse(raw) as WideEvent;
    },
  };
}

function fixture(name: string): WideEvent {
  return JSON.parse(readFileSync(join(fixtureDir, name), "utf8")) as WideEvent;
}

function masked(ev: WideEvent): unknown {
  const v = JSON.parse(JSON.stringify(ev)) as any;
  for (const k of ["event_id", "trace_id", "span_id", "parent_span_id", "ts_start", "ts_end"]) {
    if (k in v) v[k] = v[k] === "" ? "__MASKED_EMPTY__" : "__MASKED__";
  }
  v.duration_ms = 0;
  for (const s of v.steps ?? []) {
    if ("span_id" in s) s.span_id = s.span_id === "" ? "__MASKED_EMPTY__" : "__MASKED__";
    s.start_ms = 0;
    s.duration_ms = 0;
  }
  for (const l of v.logs ?? []) l.ts_offset_ms = 0;
  return normalize(v);
}

function normalize(v: unknown): unknown {
  if (Array.isArray(v)) return v.length === 0 ? undefined : v.map(normalize).filter((x) => x !== undefined);
  if (v && typeof v === "object") {
    const out: Record<string, unknown> = {};
    for (const [k, val] of Object.entries(v)) {
      const n = normalize(val);
      if (n !== undefined && n !== "") out[k] = n;
    }
    return Object.keys(out).length === 0 ? undefined : out;
  }
  return v;
}

function expectParity(got: WideEvent, name: string): void {
  expect(masked(got)).toEqual(masked(fixture(name)));
}

describe("TS fixture parity", () => {
  it("ok-simple", async () => {
    const h = capture();
    const ctx = begin({}, { traceId: "1".repeat(32), spanId: "1".repeat(16) });
    setField(ctx, "http", { method: "POST", route: "/checkout", status: 200 });
    stepSync(ctx, "db.load_cart", () => undefined);
    await finalize(ctx);
    expectParity(h.event(), "ok-simple.json");
  });

  it("error-payment-cascade", async () => {
    const h = capture();
    const ctx = begin({}, { traceId: "2".repeat(32), spanId: "2".repeat(16) });
    setField(ctx, "http", { method: "POST", route: "/checkout", status: 200 });
    setField(ctx, "user", { id: "u_123" });
    stepSync(ctx, "db.load_cart", () => undefined);
    const err = newError("PMT_502", { reason: "upstream gateway 5xx" });
    try {
      stepSync(ctx, "payment.charge", () => {
        from(ctx).warn("retrying payment");
        recordOutgoingSpan(ctx, "3".repeat(16), "payment", "POST /charge");
        fail(ctx, err);
        throw err;
      });
    } catch {
      // Expected by the fixture driver.
    }
    setHTTPStatus(ctx, 502);
    await finalize(ctx);
    expectParity(h.event(), "error-payment-cascade.json");
  });

  it("error-panic", async () => {
    const h = capture();
    const ctx = begin({}, { traceId: "6".repeat(32), spanId: "7".repeat(16) });
    setField(ctx, "http", { method: "POST", route: "/checkout", status: 500 });
    await finalizePanic(ctx);
    expectParity(h.event(), "error-panic.json");
  });

  it("suppressed-healthcheck", async () => {
    const h = capture();
    const ctx = begin({}, { traceId: "5".repeat(32), spanId: "6".repeat(16) });
    setField(ctx, "http", { method: "GET", route: "/healthz", status: 200 });
    suppress(ctx);
    await finalize(ctx);
    expectParity(h.event(), "suppressed-healthcheck.json");
  });

  it("aborted-cancel", async () => {
    const h = capture();
    const ctx = begin({}, { traceId: "4".repeat(32), spanId: "5".repeat(16) });
    setField(ctx, "http", { method: "POST", route: "/checkout" });
    await finalizeAborted(ctx);
    expectParity(h.event(), "aborted-cancel.json");
  });

  it("timeout-watchdog", async () => {
    const h = capture();
    const ctx = begin({}, { traceId: "3".repeat(32), spanId: "4".repeat(16) });
    setField(ctx, "http", { method: "POST", route: "/checkout" });
    stepSync(ctx, "db.load_cart", () => undefined);
    stepSync(ctx, "payment.charge", () => {
      void finalizeTimeout(ctx);
    });
    expectParity(h.event(), "timeout-watchdog.json");
  });
});
